mlir/lib/Dialect/Rock/Transforms/AlignTiling.cpp

```diff
@@ -346,115 +274,137 @@ LogicalResult MILARewritePattern::matchAndRewrite(linalg::GenericOp laGeneric,
-    Value sliceVals = b.create<InBoundsLoadOp>(loc, typeToCopy, gemmOuts,
-                                               gemmStoreOp.getSourceCoord());
-    b.create<InBoundsStoreOp>(loc, sliceVals, fusionSlice, zero);
-
-    // 2.2. Tile linalg.generic with vgpr as input, return output vgprs
-    Value laOutRegs =
-        reconfigureLAGeneric(b, laGeneric, fusionSlice, idxMaps, gemmStoreOp);
-    // 2.2.0. Move the generic before the write-back. This'll put all
-    // the copy loops for other inputs before the generic due to insertion
-    // order.
-    laGeneric->moveBefore(gemmStoreOp);
-
-    // 2.3. Replace twcopy inputs with la.generic result vgprs
-
-    // Since the threadwise copy arg has gone through untransform()
-    // its expected output type is the same as the output type of the
-    // linalg.generic.
-    gemmStoreOp.getSourceMutable().assign(laOutRegs);
-    // The indexing has been moved into slice creation, reset source
-    // coord.
-    gemmStoreOp.getSourceCoordMutable().assign(zero);
-    gemmStoreOp.getDestMutable().assign(out);
+    fusionRegs = b.create<GpuAllocOp>(loc, gemmOutType);
+  }
+
+  PatternRewriter::InsertionGuard guard(b);
+  // 2.0. Reset insertion point to before the copy.
+  b.setInsertionPoint(gemmStoreOp);
+
+  // 2.2. Tile linalg.generic with vgpr as input, return output vgprs
+  Value laOutRegs = reconfigureLAGeneric(b, laGeneric, fusionRegs, gemmStoreOp);
+  // 2.2.0. Move the generic before the write-back. This'll put all
+  // the copy loops for other inputs before the generic due to insertion
+  // order.
+  laGeneric->moveBefore(gemmStoreOp);
+
+  gemmOut.replaceAllUsesWith(fusionRegs);
+
+  // 2.3. Replace rock.threadwise_write_all inputs with la.generic result vgprs
+  gemmStoreOp.setOperand(0, laOutRegs);
+
+  out = transformAccordingly(b, out, gemmStoreOp.getOperand(1));
+  gemmStoreOp.setOperand(1, out);
+
+  if (auto outAlloc = out.getDefiningOp<memref::AllocOp>()) {
+    outAlloc->moveBefore(gemmStoreOp);
+  }
 
+  return success();
+}
+
+LogicalResult MemcpyRewritePattern::matchAndRewrite(memref::CopyOp copy,
+                                                    PatternRewriter &b) const {
+  auto src = copy.getSource();
+  auto trg = copy.getTarget();
+
+  Operation *gemmStoreOp = nullptr;
+  if (auto allocOp = src.getDefiningOp<memref::AllocOp>()) {
+    if (auto twop = traceToThreadwiseWrite(src)) {
+      gemmStoreOp = twop;
+    }
+  }
+
+  if (gemmStoreOp && isa<rock::ThreadwiseWriteAllOp>(gemmStoreOp)) {
+    PatternRewriter::InsertionGuard guard(b);
+    b.setInsertionPoint(gemmStoreOp);
+
+    // 1. replace memref.copy with rock.threadwise_write_all
+    trg = transformAccordingly(b, trg, gemmStoreOp->getOperand(1));
+    gemmStoreOp->setOperand(1, trg);
+
+    if (auto outAlloc = trg.getDefiningOp<memref::AllocOp>())
+      outAlloc->moveBefore(gemmStoreOp);
+
+    b.eraseOp(copy);
     return success();
```

Comment:
Here also, shouldn't this be a pass failure and not a match failure if there memref::Copy is to be found at this stage ?

Reply:
I think so too.. but why do we need to do that there since we have it here ?

Reply:
Caught by alloc check at the end but it is kind of clumsy..